File: src/main/java/sensormodels/BluetoothSensorData.java
package sensormodels;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import java.io.File;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Date;
import org.apache.lucene.document.Document;
import org.bson.codecs.pojo.annotations.BsonIgnore;
import utils.WebAppConstants;

public class BluetoothSensorData extends DatabaseModel {

  @BsonIgnore public static final String MY_SQL_TABLE_NAME = "BluetoothSensorData";

  @BsonIgnore public static final String FILE_NAME = "Bluetooth";

  @SerializedName("sensor_name")
  @Expose
  private String sensorName;

  @SerializedName("timestamp")
  @Expose
  private String timestamp;

  @SerializedName("sensor_data")
  @Expose
  private SensorData sensorData;

  @Expose
  @SerializedName("formatted_date")
  private String formattedDate;

  private File file;

  public String getSensorName() {
    return sensorName;
  }

  public void setSensorName(String sensorName) {
    this.sensorName = sensorName;
  }

  public String getTimestamp() {
    return timestamp;
  }

  public void setTimestamp(String timestamp) {
    this.timestamp = timestamp;
  }

  public SensorData getSensorData() {
    return sensorData;
  }

  public void setSensorData(SensorData sensorData) {
    this.sensorData = sensorData;
  }

  public void setFormattedDate() {
    this.formattedDate = WebAppConstants.inputDateFormat.format(new Date(timestamp));
  }

  @Override
  public String getStartTime() {
    return this.getTimestamp();
  }

  public String getFormattedDate() {
    return formattedDate;
  }

  @Override
  @BsonIgnore
  public String getMongoCollectionName() {
    return "BluetoothSensorData";
  }

  @Override
  @BsonIgnore
  public Class<BluetoothSensorData> getClassObject() {
    return BluetoothSensorData.class;
  }

  @Override
  @BsonIgnore
  public String getTableName() {
    return MY_SQL_TABLE_NAME;
  }

  @Override
  @BsonIgnore
  public String getCreateTableQuery() {
    return "CREATE TABLE "
        + this.getTableName()
        + "(timestamp VARCHAR(30) , "
        + " formatted_date VARCHAR(10) , "
        + " sensor_name VARCHAR(30) , "
        + " state CHAR (225)) ";
  }

  @Override
  @BsonIgnore
  public String getInsertIntoTableQuery() {
    return " insert into "
        + this.getTableName()
        + " (timestamp,formatted_date,sensor_name,state)"
        + " values (?, ?, ?, ?)";
  }

  @Override
  @BsonIgnore
  public void fillQueryData(PreparedStatement preparedStmt) throws SQLException {
    preparedStmt.setString(1, this.getTimestamp());
    preparedStmt.setString(2, this.getFormattedDate());
    preparedStmt.setString(3, this.getSensorName());
    preparedStmt.setString(4, this.getSensorData().getState());
  }

  @Override
  @BsonIgnore
  public String getFileName() {
    return FILE_NAME;
  }

  @Override
  @BsonIgnore
  public File getFile() {
    return this.file;
  }

  @Override
  @BsonIgnore
  public void setFile(File file) {
    this.file = file;
  }

  @Override
  @BsonIgnore
  public Document getDocument() {
    return new Document();
  }

  public static class SensorData {
    @SerializedName("state")
    @Expose
    private String state;

    public String getState() {
      return state;
    }

    public void setState(String state) {
      this.state = state;
    }
  }
}

File: src/main/java/sensormodels/BluetoothSensorData.kt
package sensormodels

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import org.apache.lucene.document.Document
import org.bson.codecs.pojo.annotations.BsonIgnore
import sensormodels.store.models.SuperStoreModel
import utils.WebAppConstants
import java.io.File
import java.sql.PreparedStatement
import java.util.*

data class BluetoothSensorData(
    @SerializedName("sensor_name")
    @Expose
    var sensorName: String? = null,
    @SerializedName("timestamp")
    @Expose
    var timestamp: String? = null,
    @SerializedName("sensor_data")
    @Expose
    var sensorData: SensorData? = null
) : SuperStoreModel {

    @Expose
    @SerializedName("formatted_date")
    var formattedDate: String? = null
        private set
    override val fileName = FILE_NAME
    override var file: File? = null

    override fun setFormattedDate() {
        formattedDate = WebAppConstants.inputDateFormat.format(Date(timestamp))
    }

    override val startTime: String? = timestamp

    data class SensorData(
        @SerializedName("state")
        @Expose
        var state: String? = null
    )

    companion object {
        @BsonIgnore
        val MY_SQL_TABLE_NAME = "BluetoothSensorData"

        @BsonIgnore
        val FILE_NAME = "Bluetooth"
    }

    override val document = Document()
    override val mongoCollectionName = "BluetoothSensorData"
    override val tableName: String = MY_SQL_TABLE_NAME
    override val createTableQuery =
        "CREATE TABLE ${this.tableName}(timestamp VARCHAR(30) ,  formatted_date VARCHAR(10) ,  sensor_name VARCHAR(30) ,  state CHAR (225)) "
    override val insertIntoTableQuery: String =
        " insert into ${this.tableName} (timestamp,formatted_date,sensor_name,state) values (?, ?, ?, ?)"

    override fun fillQueryData(preparedStmt: PreparedStatement) {
        preparedStmt.apply {
            setString(1, timestamp)
            setString(2, formattedDate)
            setString(3, sensorName)
            setString(4, sensorData!!.state)
        }
    }
}